booking_tour_project/app/Services/ChartService.php

```diff
@@ -0,0 +1,130 @@
+            ->first();
+        $data['number_booking'] = $bookings->number_booking;
+        $data['revenue'] = $revenue ? $revenue->revenue : 0;
+        $data['register'] = $register ? $register->register : 0;
+        return $data;
+    }
+
+    private function filterYear($now)
+    {
+        $submonth = Carbon::now()->subMonths(12);
+        $date=[];
+        $data = [];
+        do{
+            array_push($date, $this->formatMonth($submonth));
+            $value = Booking::filterYear($submonth->year, $submonth->month);
+            $submonth->addMonth();
+            array_push($data, $value);
+        }while($submonth->lte($now));
+        $result['data'] = $data;
+        $result['date'] = $date;
+        return $result;
+    }
+
+    private function filterThisMonth($now)
+    {
+        $this_month = Carbon::now()->startOfMonth();
+        $result = $this->filterFunction($this_month, $now);
+        return $result;
+    }
+
+    private function filterLastMonth()
+    {
+        $early_last_month = Carbon::now()->subMonth()->startOfMonth();
+        $end_of_last_month = Carbon::now()->subMonth()->endOfMonth();
+        $result = $this->filterFunction($early_last_month, $end_of_last_month);
+        return $result;
+    }
+
+    private function filterDay($now)
+    {
+        $sub7days = Carbon::now()->subDays(7);
+        $result = $this->filterFunction($sub7days, $now);
+        return $result;
+    }
+
+    public function filterBy($input)
+    {
+        $now = Carbon::now();
+        if ($input['filter'] == 'day') {
+            $result = $this->filterDay($now);
+        } elseif ($input['filter'] == 'lastmonth') {
+            $result = $this->filterLastMonth();
+        } elseif ($input['filter'] == 'thismonth') {
+            $result = $this->filterThisMonth($now);
+        } else {
+            $result = $this->filterYear($now);
+        }
+        return $result;
+    }
+
+    public function filterDate(Request $request)
+    {
+        $dateTo = Carbon::parse($request->dateTo);
+        $dateFrom = Carbon::parse($request->dateFrom);
+        $result = $this->filterFunction($dateFrom, $dateTo);
+        return $result;
+    }
+
+    public function chartBar()
+    {
+        $value = Booking::select(DB::raw('SUM(total) as revenue'))
+            ->where('booking_date', '>', Carbon::now()->subDays(30))
+            ->groupBy(DB::raw('DAY(booking_date)'))
+            ->get();
+        $date = Booking::select('booking_date')
+            ->where('booking_date', '>', Carbon::now()->subDays(30))
+            ->groupBy('booking_date')
+            ->get();
+        $data['value'] = $value;
+        $data['date'] = $date;
+        return $data;
+    }
+
+    private function formatDate($input){
+        return $input->format('Y-m-d');
+    }
+    private function formatMonth($input){
+        return $input->format('Y-m');
+    }
+    private function filterFunction($datestart, $dateend){
```

Comment:
Tên biến nếu đặt sát nhau thì ký tự đầu của chữ thứ 2 in hoa lên e: $dateStart, $dateEnd.
Không thì mỗi thêm dấu _ giữa mỗi chữ